### PR TITLE
Have `get_param` (optionally) return samples grouped by MCMC chain.

### PR DESCRIPTION
**Pull request by null-a**
This extends the `get_param` function to include a `preserve_chains` argument. This makes it possible to extract samples grouped by MCMC chain from a `fit` object.

The motivation for this is to make it possible to compute MCMC diagnostics. After this change is merged, we'll be able to do things like:

```python
import pandas as pd
import numpy as np
from brmp import defm
from brmp.fit import get_param
from numpyro.diagnostics import effective_sample_size, gelman_rubin

df = pd.DataFrame(dict(
    y=np.random.randn(100),
))

fit = defm('y ~ 1', df).fit(iter=100, num_chains=2)

print(effective_sample_size(get_param(fit, 'b', True)))
print(gelman_rubin(get_param(fit, 'b', True)))
```

### CODE REVIEW

**Comment by null-a** [comment]
@neerajprad: Would you like to review this, or would you rather we merge now and review later, as part of a wider code review?

**Comment by neerajprad** [comment]
> @neerajprad: Would you like to review this, or would you rather we merge now and review later, as part of a wider code review?

I'll be happy to! I think limited scope changes are good opportunities to familiarize myself better with the code base.

File: brmp/backend.py [hunk 1]
```diff
@@ -113,3 +116,22 @@ def apply_default_hmc_args(iter, warmup, num_chains):
     warmup = iter // 2 if warmup is None else warmup
     num_chains = 1 if num_chains is None else num_chains
     return iter, warmup, num_chains
+
+# Helpers for working with arrays of samples. These work with both
+# numpy and torch arrays.
+
+# (num_chains, num_samples, ...) -> (num_chains * num_samples, ...)
+def flatten(arr):
+    return arr.reshape((arr.shape[0] * arr.shape[1],) + arr.shape[2:])
```
Comment (neerajprad) [inline brmp/backend.py]: nit: can be simplifed to `arr.reshape((-1,) + arr.shape[2:])`.

File: brmp/numpyro_backend.py [hunk 2]
```diff
@@ -39,15 +40,22 @@ def from_numpy(data):
 
 # TODO: Better name.
 def run_model_on_samples_and_data(modelfn, samples, data):
-    return vmap(lambda sample: handler.substitute(modelfn, sample)(**data, mode='prior_and_mu'))(samples)
+    assert type(samples) == dict
+    assert len(samples) > 0
+    num_chains, num_samples = next(iter(samples.values())).shape[0:2]
+    assert all(arr.shape[0:2] == (num_chains, num_samples) for arr in samples.values())
+    flat_samples = {k: flatten(arr) for k, arr in samples.items()}
+    out = vmap(lambda sample: handler.substitute(modelfn, sample)(**data, mode='prior_and_mu'))(flat_samples)
```
Comment (neerajprad) [inline brmp/numpyro_backend.py]: What do the the different modes do - in particular, `prior_and_mu`, `prior` and `full`? It seems to me that this will generate samples from the posterior predictive, but I am not sure if I fully follow the distinction between the three modes. I think that you would also want to `vmap` over the rng seeds since this will use the same random stream for each row. NumPyro will have an [infer_util.predictive](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer_util.py#L293) function in the next release that you may find useful. cc. @fehiepsi

File: brmp/pyro_backend.py [hunk 3]
```diff
@@ -119,24 +127,36 @@ def nuts(data, model, iter=None, warmup=None, num_chains=None):
     nuts_kernel = NUTS(model.fn, jit_compile=False, adapt_step_size=True)
     mcmc = MCMC(nuts_kernel, num_samples=iter, warmup_steps=warmup, num_chains=num_chains)
     mcmc.run(**data)
-    samples = mcmc.get_samples()
+
+    samples = mcmc.get_samples(group_by_chain=True)
+    # Pyro doesn't insert a chain dim when num_chains==1.
```
Comment (neerajprad) [inline brmp/pyro_backend.py]: We should do this in Pyro too. At one point, we didn't have the `group_by_chain` keyword argument in Pyro, and it made sense to not insert a dim for the default `num_chains=1`, but now that we have this argument which is set to False by default, we should not make any distinctions for `num_chains=1`.

File: brmp/backend.py [hunk 4]
```diff
@@ -113,3 +116,22 @@ def apply_default_hmc_args(iter, warmup, num_chains):
...
+
+# (num_chains * num_samples, ...) -> (num_chains, num_samples, ...)
+def unflatten(arr, num_chains, num_samples):
+    return arr.reshape((num_chains, num_samples) + arr.shape[1:])
+
+def main():
+    import torch
+    t = torch.rand(2, 10, 3)
```
Comment (neerajprad) [inline brmp/backend.py]: nit: Perhaps, move this to a simple test?

File: brmp/pyro_backend.py (hunk 3 above)
Comment (neerajprad) [reply brmp/pyro_backend.py]: I will fix this shortly in Pyro. If you notice any such odd differences between the two, let us know!

File: brmp/numpyro_backend.py (hunk 2 above)
Comment (fehiepsi) [reply brmp/numpyro_backend.py]: I think that we just need to use @neerajprad's trick: `numpyro.sample("mu", dist.Delta(mu), obs=mu)` and `predictive` utility for these purposes. Yesterday I tried to port chapters 4 & 5 of Statistical Rethinking book to see limitations of NumPyro API and found that that trick is pretty handy. I think what you want to achieve is something similar to code blocks 5.3, 5.4, 5.5 in [this gist](https://nbviewer.jupyter.org/gist/fehiepsi/195e080f7513f34402fefa82b16e7248#Code-5.3)?

File: brmp/backend.py (hunk 1 above)
Comment (null-a) [reply brmp/backend.py]: I thought that originally too, but in this case `arr` can have shape e.g. `(1,1,0)`, in which case reshaping to `(-1,0)` is ambiguous.

File: brmp/numpyro_backend.py (hunk 2 above)
Comment (null-a) [reply brmp/numpyro_backend.py]: > What do the the different modes do

The `mode` flag makes it possible to skip parts of the generated model when it is run. Something like:

||`prior_only`|`prior_and_mu`|`full`|
|---|---|---|---|
|Sample latents?| Y | Y | Y |
|Compute `mu`?| N | Y | Y |
|Observe `y`?| N | N | Y |

(Where `mu` is the value computed for the location of the response distribution before applying any inv. link fn.)

In case the question was more about where this is used...

The `prior_only` mode is used when running the model in order to collect samples from the prior, for example.

`prior_and_mu` is used to e.g. implement the [`location`](https://github.com/pyro-ppl/brmp/blob/11f224f693ad55851429d914178a85c369a35e2e/brmp/backend.py#L32-L40) method required of a back end. (Which is used for e.g. sampling the posterior predictive as you mentioned.) Internally, this makes use of `run_model_on_samples_and_data`, which replays a model on a bag of samples at the given data set in order to compute a `mu` vector. The `prior_and_mu` flag avoids us having to provide a bunch of observations to the model in order to do so.

(It's worth mentioning the pyro back end has a similar `mode` flag, which finds additional uses in the implementation of the `svi` method.)

Does that help? I'm happy to try to explain more carefully if not.

>  I think that you would also want to vmap over the rng seeds since this will use the same random stream for each row.

`run_model_on_samples_and_data` itself is deterministic. (Taking samples and the data as input, and computing only `mu`.) Sampling from the posterior predictive happens elsewhere. More specifically, it happens in the implementation of the [`sample_response_fn`](https://github.com/pyro-ppl/brmp/blob/11f224f693ad55851429d914178a85c369a35e2e/brmp/backend.py#L86-L89) required of a back end. For the numpyro backend, this is seeded [here](https://github.com/pyro-ppl/brmp/blob/11f224f693ad55851429d914178a85c369a35e2e/brmp/numpyro_codegen.py#L363).

Does that makes sense?

File: brmp/backend.py (hunk 4 above)
Comment (null-a) [reply brmp/backend.py]: Thanks, I'll remove this.

**Comment by null-a** [comment]
@neerajprad: Thanks for reviewing. This is now ready for you to take another look.

> you can probably just dispatch these to NumPyro's diagnostics module since it should work with standard numpy arrays.

Exactly. I also intend to follow-up with a PR that uses the functionality added here to add `n_eff` and `r_hat` columns to the output of `brmp.fit.marginals`.

File: brmp/backend.py (hunk 1 above)
Comment (neerajprad) [reply brmp/backend.py]: Ahh..interesting, I didn't realize that. Thanks for explaining.

File: brmp/numpyro_backend.py (hunk 2 above)
Comment (neerajprad) [reply brmp/numpyro_backend.py]: I see, that makes sense, and we do not need any seeding to happen in that case. Thanks for explaining it so well! I will take another look at how the generated code looks like for the three modes that you mentioned. I think @fehiepsi's suggestion might be applicable there, but in any case, that's more of an internal detail and it doesn't matter how we record the value of `mu`.

**Comment by null-a** [comment]
Great, thanks for reviewing and merging.

> let me collate a few of them and ask you that separately

Sure, sounds good.
